pdr_backend/pdr_dashboard/callbacks/callbacks_feeds.py

```diff
@@ -0,0 +1,166 @@
+import dash
+from dash import Input, Output, State, callback_context
+
+
+def filter_table_by_range(min_val, max_val, label_text):
+    ctx = callback_context
+    button_id = ctx.triggered[0]["prop_id"].split(".")[0]
+    if (not min_val and not max_val) or button_id == "clear_filters_button":
+        return label_text
+
+    return f"{label_text} {min_val if min_val else ''}-{max_val if max_val else ''}"
+
+
+def filter_condition(item, field, values):
+    if not values:
+        return True
+    return item[field] in values
+
+
+def range_condition(item, field, min_value, max_value):
+    if min_value is not None and min_value != "" and item[field] < min_value:
+        return False
+    if max_value is not None and max_value != "" and item[field] > max_value:
+        return False
+    return True
+
+
+def get_callbacks_feeds(app):
+    @app.callback(
+        Output("feeds_page_table", "data"),
+        [
+            Input("base_token", "value"),
+            Input("quote_token", "value"),
+            Input("venue", "value"),
+            Input("time", "value"),
+            Input("sales_button", "n_clicks"),
+            Input("revenue_button", "n_clicks"),
+            Input("accuracy_button", "n_clicks"),
+            Input("volume_button", "n_clicks"),
+        ],
+        State("sales_min", "value"),
+        State("sales_max", "value"),
+        State("revenue_min", "value"),
+        State("revenue_max", "value"),
+        State("accuracy_min", "value"),
+        State("accuracy_max", "value"),
+        State("volume_min", "value"),
+        State("volume_max", "value"),
+    )
+    def filter_table(
+        base_token,
+        quote_token,
+        venue,
+        time,
+        _n_clicks_sales,
+        _n_clicks_revenue,
+        _n_clicks_accuracy,
+        _n_clicks_volume,
+        sales_min,
+        sales_max,
+        revenue_min,
+        revenue_max,
+        accuracy_min,
+        accuracy_max,
+        volume_min,
+        volume_max,
+    ):
+        """
+        Filter table based on selected dropdown values.
+        """
+
+        new_table_data = [
+            item
+            for item in app.feeds_table_data
+            if filter_condition(item, "base_token", base_token)
+            and filter_condition(item, "quote_token", quote_token)
+            and filter_condition(item, "venue", venue)
+            and filter_condition(item, "time", time)
+            and range_condition(item, "sales", sales_min, sales_max)
+            and range_condition(item, "sales_revenue_(OCEAN)", revenue_min, revenue_max)
+            and range_condition(item, "avg_accuracy", accuracy_min, accuracy_max)
+            and range_condition(item, "volume_(OCEAN)", volume_min, volume_max)
```

Comment:
I feel like there are lots of "and" in this condition.
Can we do like that ?


```

conditions = [
            ("filter", "base_token", base_token),
            ("filter", "quote_token", quote_token),
            ("filter", "venue", venue),
            ("filter", "time", time),
            ("range", "sales", sales_min, sales_max),
            ("range", "sales_revenue_(OCEAN)", revenue_min, revenue_max),
            ("range", "avg_accuracy", accuracy_min, accuracy_max),
            ("range", "volume_(OCEAN)", volume_min, volume_max),
        ]

        def check_condition(item, condition_type, field, *values):
            if condition_type == "filter":
                return filter_condition(item, field, values[0])
            elif condition_type == "range":
                return range_condition(item, field, values[0], values[1])
            return True

        new_table_data = [
            item
            for item in app.feeds_table_data
            if all(check_condition(item, *condition) for condition in conditions)
        ]

        return new_table_data
```

Reply:
This approach gives errors for some reason

Reply:
Discovered it was something else

Reply:
Thanks for implementation